Python/optimizer.c

```diff
@@ -0,0 +1,248 @@
+
+#include "Python.h"
+#include "opcode.h"
+#include "pycore_interp.h"
+#include "pycore_opcode.h"
+#include "pycore_pystate.h"
+#include "cpython/optimizer.h"
+#include <stdbool.h>
+#include <stdint.h>
+#include <stddef.h>
+
+static int32_t
+get_next_free_in_executor_array(PyCodeObject *code)
+{
+    _PyExecutorArray *old = code->co_executors;
+    int size = 0;
+    int capacity = 0;
+    if (old != NULL) {
+        size = old->size;
+        capacity = old->capacity;
+        if (capacity >= 256) {
+            PyErr_Format(PyExc_SystemError,
+                        "too many executors for code object");
+            return -1;
+        }
+    }
+    for (uintptr_t i = 0; i < size; i++) {
+        if (old->executors[i] == NULL) {
+            return i;
+        }
+    }
+    if (size >= capacity) {
+        int new_capacity = capacity ? capacity * 2 : 4;
+        _PyExecutorArray *new = PyMem_Malloc(offsetof(_PyExecutorArray, executors) +
+            new_capacity * sizeof(_PyExecutorObject *));
+        if (new == NULL) {
+            PyErr_NoMemory();
+            return -1;
+        }
+        code->co_executors = new;
+        new->capacity = new_capacity;
+        for (uintptr_t i = 0; i < size; i++) {
+            new->executors[i] = old->executors[i];
+        }
+        if (old != NULL) {
+            PyMem_Free(old);
+        }
+    }
+    assert(size < code->co_executors->capacity);
+    code->co_executors->size = size + 1;
+    return size;
+}
+
+static int
+insert_executor(PyCodeObject *code, int offset, _PyExecutorObject *executor)
+{
+    _Py_CODEUNIT original = _PyCode_CODE(code)[offset];
+    if (original.op.code == ENTER_EXECUTOR) {
+        _PyExecutorObject *old = code->co_executors->executors[original.op.arg];
+        executor->vm_data.opcode = old->vm_data.opcode;
+        executor->vm_data.oparg = old->vm_data.oparg;
+        old->vm_data.opcode = 0;
+        code->co_executors->executors[original.op.arg] = executor;
+        Py_DECREF(old);
+    }
+    else {
+        int32_t index = get_next_free_in_executor_array(code);
+        if (index < 0) {
+            return -1;
+        }
+        Py_INCREF(executor);
+        executor->vm_data.opcode = original.op.code;
+        executor->vm_data.oparg = original.op.arg;
+        code->co_executors->executors[index] = executor;
+        assert(index < 256);
+        _PyCode_CODE(code)[offset].op.code = ENTER_EXECUTOR;
+        _PyCode_CODE(code)[offset].op.arg = index;
+    }
+    return 0;
+}
+
+int PyUnstable_Replace_Executor(PyCodeObject *code, int offset, _PyExecutorObject *new)
```

Comment:
Could you document above this function that it takes a borrowed reference to an executor object please?

Reply:
Isn't that the default? I thought we only documented non-default behaviors.

Reply:
Oh, it consumes a reference. It shouldn't. I'll fix that.